### PR TITLE
context_drm_egl: Use gbm_surface_create_with_modifiers

### PR DESCRIPTION
**Pull request by philipl**
The GBM supporting nvidia driver doesn't support creating surfaces
without modifiers and using modifiers is more and more recommended as
the right way to do this.

Enumerating modifiers is painfully verbose, but necessary if we are to
allow the driver to pick the best possible one.

Finally, I added support for [A|X]BGR2101010 because nvidia supports
that instead of RGB. Of course, their EGL doesn't support 10bit formats
so you still can't actually use a 10bit DRM format.

### CODE REVIEW

**Comment by emersion** [comment]
> This change just always uses MOD_LINEAR which is probably suboptimal
on other hardware where the default format used without modifiers
might be a more efficient tiled format.

Yes, correct. The right way to fix this is to grab the possible modifiers from the `IN_FORMATS` KMS plane prop.

**Comment by philipl** [comment]
Updated to enumerate modifiers and support the BGR 10bit formats that nvidia support in DRM.

**Comment by emersion** [comment]
Some drivers support atomic, but don't support format modifiers (e.g. GFX8- on amdgpu). I don't remember whether a manual fallback to `gbm_surface_create` is required, testing to make sure that code-path isn't regressed would be nice.

For the modifier-aware drivers, you'll need to use `AddFB2WithModifiers` instead of `AddFB2`, or risk having the driver interpret the buffer as linear when it's tiled.

**Comment by philipl** [comment]
Updated to use `AddFB2WithModifiers`. I don't have any access to hardware that doesn't support modifiers. If anyone does and could try this out, it would be great. I'd like to avoid adding the fallback code if it's avoidable.

File: video/out/drm_common.c [hunk 1]
```diff
@@ -90,7 +90,8 @@ const struct m_sub_options drm_conf = {
             M_RANGE(0, INT_MAX)},
         {"drm-format", OPT_CHOICE(drm_format,
             {"xrgb8888",    DRM_OPTS_FORMAT_XRGB8888},
-            {"xrgb2101010", DRM_OPTS_FORMAT_XRGB2101010})},
+            {"xrgb2101010", DRM_OPTS_FORMAT_XRGB2101010},
+            {"xbgr2101010", DRM_OPTS_FORMAT_XBGR2101010})},
```
Comment (sfan5) [inline video/out/drm_common.c:94]: this should be in a separate commit

File: video/out/opengl/context_drm_egl.c [hunk 2]
```diff
@@ -273,15 +285,25 @@ static void update_framebuffer_from_bo(struct ra_ctx *ctx, struct gbm_bo *bo)
...
+
+    const int num_planes = gbm_bo_get_plane_count(bo);
+    for (int i = 0; i < num_planes; ++i) {
+        handles[i] = gbm_bo_get_handle_for_plane(bo, i).u32;
+        strides[i] = gbm_bo_get_stride_for_plane(bo, i);
+        offsets[i] = gbm_bo_get_offset(bo, i);
+        modifiers[i] = modifier;
+    }
+
+    int ret = drmModeAddFB2WithModifiers(fb->fd, fb->width, fb->height,
```
Comment (emersion) [inline video/out/opengl/context_drm_egl.c:303]: This call will definitely need a fallback for the `modifier == DRM_FORMAT_MOD_INVALID` case.

**Comment by philipl** [comment]
Split the BGR format addition into a separate commit and added fallback for both gbm and fb without modifiers.

File: video/out/opengl/context_drm_egl.c [hunk 3]
```diff
@@ -707,6 +750,64 @@ static bool probe_gbm_format(struct ra_ctx *ctx, uint32_t argb_format, uint32_t
         result = true;
     }
 
+    drmModeObjectProperties *props =
```
Comment (emersion) [inline video/out/opengl/context_drm_egl.c:753]: Missing `drmModeFreeObjectProperties` call to avoid leaking memory

**Comment by emersion** [comment]
Apart from this, LGTM!

**Comment by philipl** [comment]
If anyone can test on AMD that would be great. I don't have any hardware for that.

**Comment by sfan5** [comment]
Tested the current state on an (Amlogic S905X3), it picks up several modifiers and works fine.

Additional thought: what about supporting [XA]BGR8888? My board has a few compressed formats that only do bgr:
<pre>
    │       │   ├───ARM_AFBC(BLOCK_SIZE = 16x16, YTR, SPLIT, SPARSE) (0x800000000000071)
    │       │   │   ├───ABGR8888 (0x34324241)
    │       │   │   └───XBGR8888 (0x34324258)
    │       │   ├───ARM_AFBC(BLOCK_SIZE = 32x8, YTR, SPARSE) (0x800000000000052)
    │       │   │   ├───ABGR8888 (0x34324241)
    │       │   │   └───XBGR8888 (0x34324258)
    │       │   ├───ARM_AFBC(BLOCK_SIZE = 32x8, YTR, SPLIT, SPARSE) (0x800000000000072)
    │       │   │   ├───ABGR8888 (0x34324241)
    │       │   │   └───XBGR8888 (0x34324258)
</pre>

File: video/out/opengl/context_drm_egl.c [hunk 4]
```diff
@@ -707,6 +750,65 @@ static bool probe_gbm_format(struct ra_ctx *ctx, uint32_t argb_format, uint32_t
...
+                if (mod->formats & (1ull << k)) {
+                    uint32_t fmt = fmts[k + mod->offset];
+                    if (fmt == p->gbm_format) {
+                        MP_TARRAY_APPEND(p, p->gbm_modifiers,
+                                         p->num_gbm_modifiers, mod->modifier);
+                        MP_VERBOSE(ctx->vo, "Supported modifier: 0x%llX\n",
+                                   mod->modifier);
+                        break;
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (philipl) [reply video/out/opengl/context_drm_egl.c:799]: The types here are weird. the `mod->modifier` is not a `uint64_t`. It's an unsigned long long. I could cast or assign it before printing.

[the remark this answers is not in the document]
Comment (sfan5) [reply video/out/opengl/context_drm_egl.c:799]: Casting sounds okay.

File: video/out/opengl/context_drm_egl.c [hunk 5]
```diff
@@ -707,6 +750,65 @@ static bool probe_gbm_format(struct ra_ctx *ctx, uint32_t argb_format, uint32_t
...
+            }
+        }
+
+        drmModeFreePropertyBlob(blob);
+        drmModeFreeProperty(prop);
+    }
+
+    if (p->num_gbm_modifiers == 0) {
+        MP_WARN(ctx->vo, "No supported DRM modifiers found.\n");
```
[the remark this answers is not in the document]
Comment (philipl) [reply video/out/opengl/context_drm_egl.c:808]: I figured that if you claim to support modifiers and then don't actually have any compatible ones, that's a bit weird. But I can downgrade to verbose and we see what happens in reality.

File: video/out/opengl/context_drm_egl.c [hunk 6]
```diff
@@ -800,6 +909,10 @@ static bool drm_egl_init(struct ra_ctx *ctx)
         MP_ERR(ctx, "Failed to lock GBM surface.\n");
         return false;
     }
+
+    uint64_t modifier = gbm_bo_get_modifier(new_bo);
+    MP_VERBOSE(ctx, "GBM surface using modifier 0x%"PRIX64"\n", modifier);
```
[the remark this answers is not in the document]
Comment (philipl) [reply video/out/opengl/context_drm_egl.c:914]: I've moved it into `update_framebuffer_from_bo`

File: video/out/opengl/context_drm_egl.c [hunk 7]
```diff
@@ -707,6 +750,65 @@ static bool probe_gbm_format(struct ra_ctx *ctx, uint32_t argb_format, uint32_t
         result = true;
     }
 
+    drmModeObjectProperties *props =
```
[the remark this answers is not in the document]
Comment (philipl) [reply video/out/opengl/context_drm_egl.c:753]: Moved to a new function.

**Comment by philipl** [comment]
Updated to address all feedback.

File: video/out/opengl/context_drm_egl.c [hunk 8]
```diff
@@ -239,12 +242,22 @@ static bool init_gbm(struct ra_ctx *ctx)
...
-        GBM_BO_USE_SCANOUT | GBM_BO_USE_RENDERING);
+    if (p->num_gbm_modifiers == 0) {
+        p->gbm.surface = gbm_surface_create(
+            p->gbm.device,
+            p->draw_surface_size.width,
+            p->draw_surface_size.height,
+            p->gbm_format,
+            GBM_BO_USE_SCANOUT | GBM_BO_USE_RENDERING);
+    } else {
+        p->gbm.surface = gbm_surface_create_with_modifiers(
```
[the remark this answers is not in the document]
Comment (sfan5) [reply video/out/opengl/context_drm_egl.c:269]: I was wondering if we can rely on create_with_modifier to exist if `drmModeAddFB2WithModifiers` is assured to exist (already covered by current check).

File: video/out/opengl/context_drm_egl.c [hunk 9]
```diff
@@ -711,6 +763,72 @@ static bool probe_gbm_format(struct ra_ctx *ctx, uint32_t argb_format, uint32_t
...
+                    uint32_t fmt = fmts[k + mod->offset];
+                    if (fmt == p->gbm_format) {
+                        MP_TARRAY_APPEND(p, p->gbm_modifiers,
+                                         p->num_gbm_modifiers, mod->modifier);
+                        MP_VERBOSE(ctx->vo, "Supported modifier: 0x%"PRIX64"\n",
+                                   (uint64_t)mod->modifier);
+                        break;
+                    }
+                }
+            }
```
Comment (sfan5) [inline video/out/opengl/context_drm_egl.c:817]: this is still indented more than necessary (see last suggestion)

File: video/out/opengl/context_drm_egl.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (evelikov) [reply video/out/opengl/context_drm_egl.c:269]: They are different APIs provided by separate libraries, so that's a no.

The drm API mentioned is introduced with 2.4.71 - back in 2016 - and we already require newer version.

[the remark this answers is not in the document]
Comment (philipl) [reply video/out/opengl/context_drm_egl.c:269]: I've added a version check. I'm going to claim that it's reasonable to just require >= 17.1.0 than to add an #ifdef.

**Review by sfan5 — APPROVED** [review]
Works on Amlogic, though turns out there is no EGLConfig for `xbgr8888`
Still doesn't work on NVIDIA for me <i>\*shrug\*</i>
